File: src/infrastructure/coinmarketcap/CoinMarketCapClient.cs
﻿using System.Collections.Generic;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;
using core.fs.Adapters.Cryptos;
using core.Shared;
using Microsoft.Extensions.Logging;
using Microsoft.FSharp.Core;

namespace coinmarketcap
{
    public class CoinMarketCapClient : ICryptoService
    {
        private static readonly HttpClient _httpClient = new();
        private readonly ILogger<CoinMarketCapClient>? _logger;

        public CoinMarketCapClient(ILogger<CoinMarketCapClient>? logger, string accessToken)
        {
            _httpClient.DefaultRequestHeaders.Add("X-CMC_PRO_API_KEY", accessToken);
            _logger = logger;
        }

        public async Task<Listings> GetAll()
        {
            var url = "https://pro-api.coinmarketcap.com/v1/cryptocurrency/listings/latest";

            var response = await _httpClient.GetAsync(url);

            var content = await response.Content.ReadAsStringAsync();
            
            if (response.IsSuccessStatusCode == false)
            {
                throw new System.Exception("Could not get listings: " + content);
            }
            
            var value = JsonSerializer.Deserialize<Listings>(content);
            if (value == null)
            {
                throw new System.Exception("Could not deserialize response: " + content);
            }

            return value;
        }

        public async Task<FSharpOption<Datum>> Get(string token)
        {
            var prices = await GetAll();

            return prices.TryGet(token);
        }

        public async Task<Dictionary<string, Datum>> Get(IEnumerable<string> tokens)
        {
            var prices = await GetAll();

            var result = new Dictionary<string, Datum>();

            foreach (var token in tokens)
            {
                var price = prices.TryGet(token);
                if (price != null)
                {
                    result.Add(token, price.Value);
                }
                else
                {
                    _logger?.LogError("Did not find price for " + token);
                }
            }

            return result;
        }
    }
}

File: src/infrastructure/coinmarketcap.fs/CoinMarketCapClient.fs
namespace coinmarketcap

open System
open System.Collections.Generic
open System.Net.Http
open System.Text.Json
open System.Threading.Tasks
open Microsoft.Extensions.Logging
open core.fs.Adapters.Cryptos

type CoinMarketCapClient(logger: ILogger<CoinMarketCapClient> option, accessToken: string) =
    
    static let mutable httpClientInstance : HttpClient option = None
    static let lockObj = obj()
    
    let httpClient = 
        lock lockObj (fun () ->
            match httpClientInstance with
            | Some client -> client
            | None ->
                let client = new HttpClient()
                client.DefaultRequestHeaders.Add("X-CMC_PRO_API_KEY", accessToken)
                httpClientInstance <- Some client
                client
        )
    
    member _.GetAll() : Task<Listings> =
        task {
            let url = "https://pro-api.coinmarketcap.com/v1/cryptocurrency/listings/latest"
            
            use! response = httpClient.GetAsync(url)
            let! content = response.Content.ReadAsStringAsync()
            
            if not response.IsSuccessStatusCode then
                raise (Exception($"Could not get listings: {content}"))
            
            let value = JsonSerializer.Deserialize<Listings>(content)
            
            if isNull (box value) then
                raise (Exception($"Could not deserialize response: {content}"))
            
            return value
        }
    
    member this.Get(token: string) : Task<Datum option> =
        task {
            let! prices = this.GetAll()
            return prices.TryGet(token)
        }
    
    member this.Get(tokens: string seq) : Task<Dictionary<string, Datum>> =
        task {
            let! prices = this.GetAll()
            
            let result = Dictionary<string, Datum>()
            
            for token in tokens do
                match prices.TryGet(token) with
                | Some datum ->
                    result.Add(token, datum)
                | None ->
                    match logger with
                    | Some log -> log.LogError($"Did not find price for {token}")
                    | None -> ()
            
            return result
        }
    
    interface ICryptoService with
        member this.GetAll() = this.GetAll()
        member this.Get(token: string) = this.Get(token)
        member this.Get(tokens: string seq) = this.Get(tokens)
